### ISSUE TITLE
.NET8 - UseMauiEssentials not working on project targeting net6.0

### ISSUE THREAD

**Issue by projectgoav** [dotnet/maui#18748]
### Description

I updated Visual Studio 22 Windows to Version 17.8. This has installed the .NET8 SDK on my system. 

A class library that includes `<UseMauiEssentials>` does not build. 

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0103	The name 'Connectivity' does not exist in the current context
Error	CS0234	The type or namespace name 'Maui' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)
```

I am then told the solution is to install the `Xamarin.Essentials` package which is not compatible with any .NET SDK project.

![image](https://github.com/dotnet/maui/assets/6363155/d0a806b1-37ee-42cc-8008-5d4b91da14d5)

The project builds if:
- I create a `global.json` and point to .NET7 or .NET6 SDKs (which had to be reinstalled as it seems .NET8 nukes those upon install)
- I change the target project to `net7.0`

### Steps to Reproduce

1. Create a new class library targeting net6.0
2. Edit project file to look like:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
      <UseMauiEssentials>true</UseMauiEssentials>
  </PropertyGroup>
</Project>
```
3. Add the following code to `Class1.cs`
```C#
using Microsoft.Maui.Networking;
namespace MauiEssentialBug {
    public class Class1 {
        public void Foo()  {
            var x = Connectivity.NetworkAccess;
            System.Diagnostics.Debug.WriteLine($"Foo = {x}");
        }
    }
}
```
4. Fails to compile

### Link to public reproduction project repository

https://github.com/projectgoav/MauiEssentialsBug

### Version with bug

Unknown/Other

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows

### Affected platform versions

.NET8 / VS22 17.8

### Did you find any workaround?

Upgrading the project to net7.0
Adding a global.json to point to .NET7 SDK

**Comment by leonluc-dev** [comment]
It might be an unintended side effect to the fact that .NET 8 you are supposed to replace the `UseMauiEssentials` tag with a package?

This is indicated by a warning when building a .NET 8 project  `Starting with .NET 8, setting  <UseMauiEssentials>true</UseMauiEssentials>  does not automatically include NuGet package references in your project.  Update your project by including this item:  <PackageReference Include="Microsoft.Maui.Essentials" Version="8.0.3" />.  You can skip this warning by setting  <SkipValidateMauiImplicitPackageReferences>true</SkipValidateMauiImplicitPackageReferences>  in your project file.`

Does adding that package work in .NET 6 as well?

**Comment by projectgoav** [comment]
> This is indicated by a warning when building a .NET 8 project

I don't get any compiler warning at all when using the .NET8 SDK. For both the net6.0 and net7.0-ios|android projects all I get are errors like what I provided above. 

Could you point me towards any documentation / discussion about this change and why it occured?


> Does adding that package work in .NET 6 as well?

TL;DR: Adding the package works, but I have to remove `<UseMauiEssentials>` from all projects AND have to include direct references to anything we used from [PlayServiceComponents](https://github.com/xamarin/GooglePlayServicesComponents)

The net6.0 project wouldn't get me add the `MSFT.Maui.Essentials` package due to version downgrades and so I had to remove `<UseMauiEssentials>` from the csproj first. 

The net7.0-android project that included this library would then no longer build. That had the same problem reported of downgraded packages. Only solution was to remove `<UseMauiEssentials>` and then add the Nuget package manually. 

However, this then had the problem that anything from https://github.com/xamarin/GooglePlayServicesComponents was no longer included by default. Eg - We used Glide to load images in our app and so I had to reference `Xamarin.Android.Glide` directly. This had to be done for all libraries we used like this. 

After that, our app compiles and run. (I have no yet tested on iOS).

**Comment by leonluc-dev** [comment]
> I don't get any compiler warning at all when using the .NET8 SDK. For both the net6.0 and net7.0-ios|android projects all I get are errors like what I provided above.

Sorry, I was specifically referring to `net8.0` targets (`net8.0-android` etc.), not just the SDK.

Either the warning not showing up for .NET6/7 runtimes is a bug or the package needing to be explicit for .NET 6/7 runtimes (instead of just for .NET 8 runtimes) is a bug.

>Could you point me towards any documentation / discussion about this change and why it occured?

The change came from PR #14770 . Aside from that and the warning itself there is no other documentation that I know of.

**Comment by konkolyitibi** [comment]
I think I have the same issue with MAUI android, TargetFramework: net6.0-android33.0
I have no option for select android target, but the project builds when i change the target to net8.0 android

And also same ".NET6 SDKs (which had to be reinstalled as it seems .NET8 nukes those upon install"

**Comment by mattleibow** [comment]
.NET 6 is no longer supported https://dotnet.microsoft.com/en-us/platform/support/policy/maui

If you need to use .NET 6 for mobile, please use an older SDK.

**Comment by projectgoav** [comment]
Are you saying that we must change our target framework with each new .NET SDK release? Must we now target `net8(-XXX)` with all projects now that the .NET8 SDK is released, changing to `net9(-XXX)` next year when that is launched? 

Does this also mean the `netstandardX.Y` targets are also no longer supported and we must change those as well?

**Comment by w3ori** [comment]
> .NET 6 is no longer supported https://dotnet.microsoft.com/en-us/platform/support/policy/maui
> 
> 
> 
> If you need to use .NET 6 for mobile, please use an older SDK.

Then how can I use the .NET 6 SDK with Visual Studio 17.8?

I have projects in production that uses.NET 6.

**Comment by leonluc-dev** [comment]
> Are you saying that we must change our target framework with each new .NET SDK release? Must we now target `net8(-XXX)` with all projects now that the .NET8 SDK is released, changing to `net9(-XXX)` next year when that is launched?

Seems that way. Or at least every two versions if the release cadance remains similar.: MAUI 7 support will drop with the .NET 9 SDK, MAUI 8 drops with .NET 10 SDK etc.

It's similar to the Xamarin.Forms release cadance back in the day.

**Comment by projectgoav** [comment]
Argh! We're not even a MAUI app - we were using the Xamarin native SDKs. But these have been wrapped up into .NET for Android/iOS and then forced us onto the MAUI infrastructure for the Essentials and support packages. 

A bit rubbish if we've then to follow whatever the MAUI release schedule as well given Xamarin native & Xamarin.Forms used to be pretty distinct.

**Comment by leonluc-dev** [comment]
True, but native Xamarin.iOS and Xamarin.Android SDKs/IDEs dropped support beyond 2 major versions as well. So this is a similar issue in a new coating, just tied to a November release instead of the usual September one.

That said, is there anything major blocking you from migrating from `net6.0-android` to at least `net7.0-android` right now?

**Comment by mattleibow** [comment]
Unfortunately, we could not get it perfect for .NET 7 but for .NET 8 you no longer need the UseMauiEssentials and can just directly reference the https://www.nuget.org/packages/Microsoft.Maui.Essentials/8.0.3 package in your app. It supports .NET 7 and .NET 8 and hopefully when .NET 9 ships I can keep .NET 7.

If you _just_ want essentials, then you don't have to me a maui app to use the essentials and graphics nugets.

**Comment by projectgoav** [comment]
> That said, is there anything major blocking you from migrating from net6.0 to at least net7.0 targetframeworks right now?

The shared library code our apps use was converted to target `net6.0` to make use of `UseMauiEssentials` as this was the recommended upgrade path from Xamarin.Essentials. We then migrated our app projects to target `net7.0-android` and `net7.0-ios` as some of our 3rd party libraries only supporyed `net7.0` at the time. 

The conversion from legacy Xamarin native to .NET was certainly not painfree. We raised around 35 issues that we encountered across many of the .NET components. While some of these are minor (compiler warnings emitted incorrectly) there are some critical issues that have required us to develop work arounds. We also require to use an older version of Visual Studio due to an unresolved issue and I am unsure the compatibility of the older version with newer .NET SDKs. 


> If you _just_ want essentials, then you don't have to be a maui app to use the essentials and graphics nugets.

I understand this is the case now and directly adding the package and removing `UseMauiEssentials` has resolved this issue in this case. However, there has been little to no documentation of this change.

- It has been suggested that a compiler message should be generated to warn users of this. This appears not to be the case with .NET 8 building a library targeting `net6.0`.
- The Maui .NET 8 page (https://learn.microsoft.com/en-us/dotnet/maui/migration/native-projects?view=net-maui-8.0) does not mention you to use the Maui.Essentials Nuget package.
- The above links to these upgrade guides (https://devblogs.microsoft.com/xamarin/upgrade-xamarin-ios-android-to-dotnet6-dotnet7/, https://learn.microsoft.com/en-us/dotnet/maui/migration/native-essentials?view=net-maui-8.0) that still encourage you to make use of `UsesMauiEssentials`. I suggest these should be changed to direct people to install the Nuget package instead.

**Comment by thirstyape** [comment]
> .NET 6 is no longer supported https://dotnet.microsoft.com/en-us/platform/support/policy/maui
> 
> If you need to use .NET 6 for mobile, please use an older SDK.

So does this mean that .NET MAUI exists outside of the LTS support policy that applies the rest of the framework? That seems incredibly counterintuitive and shortsighted. Why even bother having an LTS edition if developers cannot rely on having their production applications work for the duration.

**Comment by leonluc-dev** [comment]
> > .NET 6 is no longer supported https://dotnet.microsoft.com/en-us/platform/support/policy/maui
> > If you need to use .NET 6 for mobile, please use an older SDK.
> 
> So does this mean that .NET MAUI exists outside of the LTS support policy that applies the rest of the framework? That seems incredibly counterintuitive and shortsighted. Why even bother having an LTS edition if developers cannot rely on having their production applications work for the duration.

It seems to be a separate support cycle altogether:

.NET support cycle: https://dotnet.microsoft.com/en-us/platform/support/policy/maui
MAUI support cycle: https://dotnet.microsoft.com/en-us/platform/support/policy/maui

Which is a bit of an issue considering .NET and MAUI version are linked in .NET 6 (unless there is a workaround to that?)

With MAUI/.NET version being separated into NuGet packages from .NET 7 onward there shouldn't be too much of an issue for the future, but it's still an issue for earlier yet still supported platforms like .NET 6...

**Comment by w3ori** [comment]
So we have to upgrade MAUI in every year, because the support cycle ends.
That not sounds too good.

Do you have any idea to use the .NET MAUI 6 SDK with Visual Studio 17.8?

I have projects in production that uses .NET MAUI 6.

I tried the new .NET MAUI 8, but I have more bugs than before, doesn't worth to upgrade.

**Comment by jamesmontemagno** [comment]
For this to work you would also need .NET MAUI workload for .NET 6 installed on your machine as well.

They need to match up. It seems like your .NET 7 install has the workloads

**Comment by projectgoav** [comment]
> For this to work you would also need .NET MAUI workload for .NET 6 installed on your machine as well.
> 
> They need to match up. It seems like your .NET 7 install has the workloads

Is this in response to the original issue? If so, then the problem still occurs with MAUI workload installed alongside the .NET8 SDK.

**Comment by jfversluis** [comment]
I feel this issue has derailed a bit and I'm unclear about the current status.

If you want to use Essentials without using .NET MAUI, you should be able to do that and docs on how can be found [here](https://learn.microsoft.com/dotnet/maui/migration/native-essentials?view=net-maui-8.0&tabs=android).

As mentioned .NET 6 (for .NET MAUI) is out of support as per the [support policy](https://dotnet.microsoft.com/platform/support/policy/maui). Reasons for that have been pointed out in the [linked issue](https://github.com/dotnet/maui/issues/19160) that was raised.

If there is still something that does not work correctly for this scenario please open a new issue with more details and let's try to stick to that topic. For this particular problem I would be interested in details information about what workloads are installed, what global.json are you using, what is the build output and non-cropped screenshots.

But I hope all is sorted by now! Thanks!

**Comment by projectgoav** [comment]
> But I hope all is sorted by now!

Nothing has been sorted and the problem still remains. I'll try and outline below.

> As mentioned .NET 6 (for .NET MAUI) is out of support

You are correct, but I don't believe there are any requirements to upgrade class library target frameworks with every net .NET SDK release. Therefore, having class libraries in projects targetting `.net6.0` _should_ still be supported, as are `netstandard2.0` and `net461` for example. 

Yet, this is no longer the case with .NET 8 if you are making use of `UseMauiEssentials`. Anything that includes `<UseMauiEssentials>` in the csproj no longer builds under .NET 6, 7 or 8 without change.

@leonluc-dev has explained that the `UseMauiEssentials` no longer automatically includes the required packages. The .NET 8 SDK apparently has a warning about this however the .NET 6 & 7 SDKs do not. 

This is a breaking change, and should be communicated as such. Maybe some developer documentation, release notes or backporting the compiler warning to older .NET SDKs? Note - the .NET6 SDK is still supported until 24th November 2024. 

> If you want to use Essentials without using .NET MAUI, you should be able to do that and docs on how can be found [here](https://learn.microsoft.com/dotnet/maui/migration/native-essentials?view=net-maui-8.0&tabs=android).

The linked information is incorrect, as are a couple of other pages. This is detailed in my [previous reply](https://github.com/dotnet/maui/issues/18748#issuecomment-1818412972).

No mention is made of the required Nuget package now that `<UseMauiEssentials`> no longer pulls in the required dependencies automatically. For those who are using the .NET 8 SDK _should_ see a compiler warning, anyone on the .NET 6 or 7 SDKs will not.


> If there is still something that does not work correctly for this scenario...

Given all the above, what is the best course of action here?

Are you suggesting I should open an issue for the breaking compilation check, another for the lack of compiler warnings on .NET 6 & 7 SDKs and a further issue for the incorrect documentation?

**Comment by thirstyape** [comment]
As noted in the links from @jfversluis, the support lifecycle for MAUI is different, .NET 6.0 MAUI was end of life on May 8, 2023.

My solution to this issue was to spin up a separate virtual machine and install the Visual Studio 17.6 LTS Build Tools. Then my build scripts work as per usual. Not ideal for interactive debug through Visual Studio, but it works. 

At the very least this has bought some time for me to gradually upgrade my projects to .NET 8.0 and have the ability to address breaking changes in my development branches.

**Comment by jfversluis** [comment]
> the .NET6 SDK is still supported until 24th November 2024.

The .NET 6 SDK is, .NET MAUI is not as already mentioned a couple of times now.

> No mention is made of the required Nuget package now that `<UseMauiEssentials>` no longer pulls in the required dependencies automatically. For those who are using the .NET 8 SDK should see a compiler warning, anyone on the .NET 6 or 7 SDKs will not.

`<UseMauiEssentials>` should still pull in the required packages on .NET MAUI 6, 7 _and_ 8. However, as of .NET 8 you have the option to get a version of Essentials (and .NET MAUI for that matter) independently of the workload that is installed on your machine.

As of .NET 8, if you do not specify a `<PackageReference ...>` for Essentials, you will get a build warning, however its just that: a warning. If you decide to ignore it, it should still work as before.

If you want to keep on using .NET 6 and/or 7 you'll have to have that workload installed on your machine, that should make this work.

My recommendation would be to get to a newer, supported version which should make this experience more smooth.

Hope that helps.

**Comment by projectgoav** [comment]
Perhaps I'm not being very clear as I struggle to grasp the naming of things post Xamarin. I'll try do below.

To start with, our project looked like this:
- Class libraries targetting `netstandard2.0`
- Class library targetting `netstandard2.0` including a reference to `Xamarin.Essentials`
- Android project targetting `Xamarin.Android` using AndroidX including a reference to `Xamarin.Essentials` 
- iOS project targetting `Xamarin.iOS` using UIKit including a reference to `Xamarin.Essentials` 

When it was announced that Xamarin was ceasing to exist we migrated to it's replacement. I think this is called `.NET for [Android|iOS]`? Our project then looked like this:
- Class libraries, targetting `netstandard2.0`
- Class library targetting `netstandard2.0` including a reference to `Xamarin.Essentials` 
- Android project targetting `.net6.0-android` using AndroidX including a reference to `Xamarin.Essentials` 
- iOS project targetting `.net6.0-ios` using UIKit including a reference to `Xamarin.Essentials` 

As part of the upgrade guide that was previously linked, we had to move from `Xamarin.Essentials` to `Maui.Essentials`. Part of that required us upgrading our `.netstandard2.0` class library to `.net6.0` as to include this package. Our project then looked like this:
- Class libraries, targetting `netstandard2.0`
- Class library targetting `.net6.0` including a reference to `Maui.Essentials` 
- Android project targetting `.net6.0-android` using AndroidX including a reference to `Maui.Essentials` 
- iOS project targetting `.net6.0-ios` using UIKit including a reference to `Maui.Essentials` 

Finally, we decided to make use of the `<UseMauiEssentials>` in the csproj files and so ended up with
- Class libraries, targetting `netstandard2.0`
- Class library targetting `.net6.0` including `<UseMauiEssentials>` 
- Android project targetting `.net6.0-android` using AndroidX including `<UseMauiEssentials>`
- iOS project targetting `.net6.0-ios` using UIKit including a reference to `<UseMauiEssentials>`

----

Can you please advise how the MAUI support cycle applies in the scenario outlined above?

**Comment by mattleibow** [comment]
The .NET MAUI support cycle is following the "Standard Term Support" or "releases happen every 12 months so the support period for STS is 18 months" : https://dotnet.microsoft.com/platform/support/policy/maui

This means that every 18 months (instead of the 3 years for LTS) you will have to update. .NET MAUI for .NET 6 was released on May 23, 2022 and thus support ended on May 8, 2023. This effectively means you should be on .NET 7 at the very least.

The main reason for this is discussed in teh link, but is really a result of the fact that supporting all the required versions of Xcode and Android SDK for many years is very hard to do correctly without breaking APIs. To get support for the required latest Xcode in .NET 6 is very expensive and not always possible.

Version | Original release date | Latest patch version | Patch release date | End of support
-- | -- | -- | -- | --
.NET MAUI 8 | November 14, 2023 | 8.0.3 | November 14, 2023 | May 14, 2025
.NET MAUI 7 | November 8, 2022 | 7.0.101 | November 7, 2023 | May 14, 2024
.NET MAUI 6 | May 23, 2022 | 6.0.553 | April 24, 2023 | May 8, 2023

**Comment by projectgoav** [comment]
To confirm: `.NET Native` projects (previously `Xamarin.Android` and `Xamarin.iOS`) must now also follow the same support cycle as MAUI?

**Comment by leonluc-dev** [comment]
> To confirm: `.NET Native` projects (previously `Xamarin.Android` and `Xamarin.iOS`) must now also follow the same support cycle as MAUI?

.NET native projects themselves (previously `Xamarin.iOS` and `Xamarin.Android`) are tied to the .NET support cycle.

It's the `Microsoft.Maui.Essentials` library (previously `Xamarin.Essentials`) that's tied to the MAUI support cycle.

If I understood everything correctly, it can be summarized as: to use Maui Essentials:

* **.NET 8 SDK with `net7.0` or `net8.0` target runtime (and onward)**
    * Use the [seperate Nuget package](https://www.nuget.org/packages/Microsoft.Maui.Essentials/8.0.7). This allows you to update the Maui Essentials package seperately from your .NET SDK, (thereby seperating the two support cycles)

* **.NET 8 SDK with `.net6.0` runtime**
    * Not possible to use Maui.Essentials. Update your target runtime to `net7.0` or up, or use SDK 6/7
 
* **.NET 6 or .NET 7 SDK**
    * Use the `<UseMauiEssentials>` property. Maui.Essentials version is linked to your .NET version so the support cycles are "linked"

**Comment by mattleibow** [comment]
I believe MAUI uses the STS cycle _because_ Android and iOS use the STS. MAUI does not care about the version of Xcode, but rather the iOS and Android projects dictate this.

The .NET SDK _only_ carries 2 versions for the iOS and Android SDKs. Current and Current-1. So, .NET 8 SDK will only have .NET 8 a d .NET 7 Android and iOS.

**Comment by davidortinau** [comment]
> To confirm: `.NET Native` projects (previously `Xamarin.Android` and `Xamarin.iOS`) must now also follow the same support cycle as MAUI?

As stated on the official support policy (https://dotnet.microsoft.com/platform/support/policy/maui) .NET MAUI encompasses all the SDKs listed. 

Also as mentioned and also noted on the official support page, .NET MAUI in .NET 6 is out-of-support, and .NET 7 will be out of support in May. This is 6 months more of support than you have been receiving with Xamarin, where a new release immediately ended the support for the previous release. https://dotnet.microsoft.com/platform/support/policy/xamarin

What @mattleibow is saying about the .NET SDK carrying 2 versions is really nothing to do with support policy. He's saying that you can build `net7.0-*` and `net8.0-*` TFMs from a single install of .NET 8. 

Whether the artifact is a NuGet or an install, the support policy is the same.

**Comment by projectgoav** [comment]
Ok, I _think_ I finally understand.

I believed that "MAUI" was essentially the replacement of Xamarin.Forms with .NET for [Platform] the replacements for Xamarin.[Platform]. 

In the Xamarin world, forms & the native stuff were pretty distinct projects and all operated on different release cycles and support systems. 

With the introduction of "MAUI", it now has wrapped up all these previously distinct technologies under the one banner, aligning all the release and support cadencies I guess?
